packages/dds/tree/src/simple-tree/prepareForInsertion.ts

```diff
@@ -0,0 +1,230 @@
+import type { ImplicitAllowedTypes, ImplicitFieldSchema } from "./schemaTypes.js";
+import { type InsertableContent, mapTreeFromNodeData } from "./toMapTree.js";
+import { UsageError } from "@fluidframework/telemetry-utils/internal";
+import { EmptyKey } from "../core/index.js";
+import { type Mutable, isReadonlyArray } from "../util/index.js";
+import {
+	getKernel,
+	type TreeNode,
+	tryUnhydratedFlexTreeNode,
+	unhydratedFlexTreeNodeToTreeNode,
+} from "./core/index.js";
+
+/**
+ * Prepare content from a user for insertion into a tree.
+ * @remarks
+ * This validates and converts the input, and if necessary invokes {@link prepareContentForHydration}.
+ */
+export function prepareForInsertion<TIn extends InsertableContent | undefined>(
+	data: TIn,
+	schema: ImplicitFieldSchema,
+	destinationContext: FlexTreeContext,
+): ExclusiveMapTree | undefined extends TIn ? undefined : never {
+	return prepareForInsertionContextless(
+		data,
+		schema,
+		getSchemaAndPolicy(destinationContext),
+		destinationContext.isHydrated() ? destinationContext : undefined,
+	);
+}
+
+/**
+ * {@link prepareForInsertion} but batched for array content.
+ * @privateRemarks
+ * TODO:
+ * Experimentally it was determined that making separate calls to prepareContentForHydration for each array item did not work.
+ * This should be understood and fixed or have the factors that cause it clearly documented.
+ * If fixed, this function should be removed, and arrays can just map over prepareForInsertion.
+ */
+export function prepareArrayForInsertion(
+	data: readonly InsertableContent[],
+	schema: ImplicitAllowedTypes,
+	destinationContext: FlexTreeContext,
+): ExclusiveMapTree[] {
+	const mapTrees: ExclusiveMapTree[] = data.map((item) =>
+		mapTreeFromNodeData(
+			item,
+			schema,
+			destinationContext.isHydrated() ? destinationContext.nodeKeyManager : undefined,
+			getSchemaAndPolicy(destinationContext),
+		),
+	);
+
+	if (destinationContext.isHydrated()) {
+		prepareContentForHydration(mapTrees, destinationContext.checkout.forest);
+	}
+
+	return mapTrees;
+}
+
+/**
+ * Split out from {@link prepareForInsertion} as to allow use without a context.
+ * @remarks
+ * Adding this entry point is a workaround for initialize not currently having a context.
+ */
+export function prepareForInsertionContextless<TIn extends InsertableContent | undefined>(
+	data: TIn,
+	schema: ImplicitFieldSchema,
+	schemaAndPolicy: SchemaAndPolicy,
+	hydratedData: Pick<FlexTreeHydratedContext, "checkout" | "nodeKeyManager"> | undefined,
+): ExclusiveMapTree | undefined extends TIn ? undefined : never {
+	const mapTree = mapTreeFromNodeData(
+		data,
+		schema,
+		hydratedData?.nodeKeyManager,
+		schemaAndPolicy,
+	);
+
+	if (mapTree !== undefined && hydratedData !== undefined) {
+		prepareContentForHydration(mapTree, hydratedData.checkout.forest);
+	}
+
+	return mapTree as ExclusiveMapTree | undefined extends TIn ? undefined : never;
+}
+
+// #region Content insertion and proxy binding
```

Comment:
Everything below here was just moved, with the only change being the removal of its `export`.